File: common/src/main/java/Segment.java
import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Optional;


record Segment(
        @JsonProperty("start") Point start,
        @JsonProperty("end") Point end
) implements Jsonisable {
    public Point derivate() {
        return end.diff(start);
    }

    public double size() {
        return end.dist(start);
    }

    public Point middle() {
        return start.add(end).div(2);
    }

    public Segment add(Point p) {
        return new Segment(start.add(p), end.add(p));
    }

    public Optional<Point> getCrossingPoint(Segment other) {
        if (size() == 0.0 || other.size() == 0.0) {
            return Optional.empty();
        }
        var diff = start.diff(other.start);
        var vA = derivate();
        var vB = other.derivate();
        var fDenom = vA.x() * vB.y() - vA.y() * vB.x();
        if (fDenom == 0.0) { // Parallel
            return Optional.empty();
        }
        var t = (diff.y() * vB.x() - diff.x() * vB.y()) / fDenom;
        var s = (diff.y() * vA.x() - diff.x() * vA.y()) / fDenom;
        if (t < 0.0f || t > 1.0f || s < 0.0f || s > 1.0f) {
            // Lines crossing outside of segment
            return Optional.empty();
        } else {
            return Optional.of(start.add(vA.mult(t)));
        }
    }
}

File: common/src/main/kotlin/demo/mesharena/common/Segment.kt
package demo.mesharena.common

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty

data class Segment @JsonCreator constructor(
        @JsonProperty("start") val start: Point,
        @JsonProperty("end") val end: Point
) : Jsonisable {

  fun derivate(): Point {
    return end.diff(start)
  }

  fun size(): Double {
    return derivate().size()
  }

  fun middle(): Point {
    return Point((start.x + end.x) / 2, (start.y + end.y) / 2)
  }

  fun add(p: Point): Segment {
    return Segment(start.add(p), end.add(p))
  }

  fun getCrossingPoint(other: Segment): Point? {
    if (size() == 0.0 || other.size() == 0.0) {
      return null
    }
    val (x, y) = start.diff(other.start)
    val vA = derivate()
    val (x1, y1) = other.derivate()
    val fDenom = vA.x * y1 - vA.y * x1
    if (fDenom == 0.0) { // Parallel
      return null
    }
    val t = (y * x1 - x * y1) / fDenom
    val s = (y * vA.x - x * vA.y) / fDenom
    return if (t < 0.0f || t > 1.0f || s < 0.0f || s > 1.0f) {
      // Lines crossing outside of segment
      null
    } else {
      start.add(vA.mult(t))
    }
  }

  override fun toString(): String {
    return "Segment{start=${start}, end=${end}}"
  }
}
